native/src/main/java/io/ballerina/stdlib/log/MaskedStringBuilder.java

```diff
@@ -0,0 +1,574 @@
+        return false;
+    }
+
+    private boolean addDefinedFieldValue(Map<String, BMap<?, ?>> fieldAnnotations, String fieldName, Object fieldValue,
+                                         boolean first) {
+        Optional<BMap<?, ?>> annotation = getLogSensitiveDataAnnotation(fieldAnnotations, fieldName);
+        Optional<String> fieldStringValue = annotation
+                .map(fieldAnnotation -> getStringValue(fieldAnnotation, fieldValue, runtime))
+                .orElseGet(() -> Optional.of(buildInternal(fieldValue)));
+
+        if (fieldStringValue.isPresent()) {
+            if (!first) {
+                stringBuilder.append(',');
+            }
+            appendFieldToJsonOptimized(fieldName, fieldStringValue.get(), annotation.isPresent(), fieldValue);
+            first = false;
+        }
+        return first;
+    }
+
+    /**
+     * Optimized version of appendFieldToJson that writes directly to StringBuilder
+     * without creating intermediate String objects for better performance.
+     */
+    private void appendFieldToJsonOptimized(String fieldName, String value, boolean hasAnnotation, Object fieldValue) {
+        stringBuilder.append('"');
+        appendEscapedStringOptimized(fieldName);
+        stringBuilder.append("\":");
+        if (hasAnnotation || fieldValue instanceof BString || fieldValue instanceof BXml) {
+            stringBuilder.append('"');
+            appendEscapedStringOptimized(value);
+            stringBuilder.append('"');
+        } else {
+            stringBuilder.append(value);
+        }
+    }
+
+    /**
+     * Optimized method to append escaped string directly to the main StringBuilder.
+     * This avoids creating intermediate String objects for better performance.
+     */
+    private void appendEscapedStringOptimized(String input) {
+        if (input == null) {
+            stringBuilder.append("null");
+            return;
+        }
+
+        if (!needsEscaping(input)) {
+            stringBuilder.append(input);
+            return;
+        }
+
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+            switch (c) {
+                case '"' -> stringBuilder.append(QUOTE_ESCAPE);
+                case '\\' -> stringBuilder.append(BACKSLASH_ESCAPE);
+                case '\b' -> stringBuilder.append(BACKSPACE_ESCAPE);
+                case '\f' -> stringBuilder.append(FORM_FEED_ESCAPE);
+                case '\n' -> stringBuilder.append(NEWLINE_ESCAPE);
+                case '\r' -> stringBuilder.append(CARRIAGE_RETURN_ESCAPE);
+                case '\t' -> stringBuilder.append(TAB_ESCAPE);
+                default -> {
+                    if (c < 0x20 || c == 0x7F) {
+                        stringBuilder.append("\\u00");
+                        stringBuilder.append(HEX_CHARS[(c >>> 4) & 0xF]);
+                        stringBuilder.append(HEX_CHARS[c & 0xF]);
```

Comment:
Shall we put a comment and/or define these [*magic values*](https://stackoverflow.com/a/47902/3615862) (0x20 and 0x7F) as constants to make it easy to understand?